extensions/amp-story/1.0/amp-story.js

```diff
@@ -1284,17 +1332,68 @@ export class AmpStory extends AMP.BaseElement {
       return Promise.resolve();
     }
 
+    // Enable paywall if required element found in DOM.
+    if (
+      isExperimentOn(this.win, 'amp-story-subscriptions') &&
+      this.element.querySelector('amp-story-subscriptions') != null &&
+      this.storeService_.get(StateProperty.SUBSCRIPTIONS_STATE) ===
+        SubscriptionsState.DISABLED
+    ) {
+      this.storeService_.dispatch(
+        Action.TOGGLE_SUBSCRIPTIONS_STATE,
+        SubscriptionsState.PENDING
+      );
+    }
+
+    const subscriptionsState = this.storeService_.get(
+      StateProperty.SUBSCRIPTIONS_STATE
+    );
+    if (
+      pageIndex >= PAYWALL_PAGE_INDEX &&
+      subscriptionsState === SubscriptionsState.PENDING
+    ) {
+      return this.blockOnPendingSubscriptionsState_(targetPageId, direction);
+    }
+
+    // Attempt to navigate to the locked pages after the paywall page.
     if (
-      isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
+      pageIndex > PAYWALL_PAGE_INDEX &&
+      subscriptionsState === SubscriptionsState.BLOCKED
     ) {
-      if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
+      this.pageAfterGranted_ = targetPageId;
+
+      // Show the paywall immediately if the active page is the paywall page.
+      if (this.isOnPaywallPage_()) {
+        this.showSubscriptionsDialog_();
         return Promise.resolve();
       }
-      this.storeService_.dispatch(Action.TOGGLE_SUBSCRIPTIONS_DIALOG, true);
 
-      // TODO(#37285): add SubscriptionService to actually trigger the subscription dialog.
+      // Redirect to the paywall page.
+      return this.switchTo_(
+        this.pages_[PAYWALL_PAGE_INDEX].element.id,
+        direction
+      );
+    }
+
+    // Attempt to navigate to the paywall page.
+    if (
+      pageIndex === PAYWALL_PAGE_INDEX &&
+      subscriptionsState === SubscriptionsState.BLOCKED
+    ) {
+      if (this.pageAfterGranted_) {
```

Comment:
I think we can get rid of this if we say show UI immediately if pageIndex > paywallIndex and just make the check

```js
if (
      pageIndex === PAYWALL_PAGE_INDEX &&
      subscriptionsState === SubscriptionsState.BLOCKED && 
      uiState !== SHOWING
    ) 
```

Reply:
Actually we don't even need the showing check because if it's the same state it would be a no-op.